File: macos/Tests/CursorTokenCoreTests/AppPathsTests.swift
import Foundation
import XCTest
@testable import CursorTokenCore

final class AppPathsTests: XCTestCase {
    func testMigratesLegacyDirectoryWhenDestMissing() throws {
        let root = FileManager.default.temporaryDirectory.appendingPathComponent("cremain-mig-\(UUID().uuidString)", isDirectory: true)
        let source = root.appendingPathComponent("CursorTokenTray", isDirectory: true)
        let dest = root.appendingPathComponent("CursorRemain", isDirectory: true)
        try FileManager.default.createDirectory(at: source, withIntermediateDirectories: true)
        try "{\"refresh_interval_minutes\":9}".write(to: source.appendingPathComponent("config.json"), atomically: true, encoding: .utf8)
        try "x\n".write(to: source.appendingPathComponent("usage_history.jsonl"), atomically: true, encoding: .utf8)
        defer { try? FileManager.default.removeItem(at: root) }

        XCTAssertTrue(AppPaths.tryMigrateLegacyDirectory(from: source, to: dest))
        XCTAssertTrue(FileManager.default.fileExists(atPath: dest.appendingPathComponent("config.json").path))
        XCTAssertTrue(FileManager.default.fileExists(atPath: dest.appendingPathComponent("usage_history.jsonl").path))
        XCTAssertFalse(FileManager.default.fileExists(atPath: source.path))
        XCTAssertFalse(AppPaths.tryMigrateLegacyDirectory(from: source, to: dest))
    }

    func testLeavesDestAloneWhenItAlreadyHasConfig() throws {
        let root = FileManager.default.temporaryDirectory.appendingPathComponent("cremain-keep-\(UUID().uuidString)", isDirectory: true)
        let source = root.appendingPathComponent("CursorTokenTray", isDirectory: true)
        let dest = root.appendingPathComponent("CursorRemain", isDirectory: true)
        try FileManager.default.createDirectory(at: source, withIntermediateDirectories: true)
        try FileManager.default.createDirectory(at: dest, withIntermediateDirectories: true)
        try "{\"from\":\"old\"}".write(to: source.appendingPathComponent("config.json"), atomically: true, encoding: .utf8)
        try "{\"from\":\"new\"}".write(to: dest.appendingPathComponent("config.json"), atomically: true, encoding: .utf8)
        defer { try? FileManager.default.removeItem(at: root) }

        XCTAssertFalse(AppPaths.tryMigrateLegacyDirectory(from: source, to: dest))
        let kept = try String(contentsOf: dest.appendingPathComponent("config.json"), encoding: .utf8)
        XCTAssertEqual(kept, "{\"from\":\"new\"}")
        XCTAssertTrue(FileManager.default.fileExists(atPath: source.appendingPathComponent("config.json").path))
    }

    func testConfigDirectoryUnderHomeMigrates() throws {
        let home = FileManager.default.temporaryDirectory.appendingPathComponent("cremain-home-\(UUID().uuidString)", isDirectory: true)
        let legacy = home.appendingPathComponent("Library/Application Support/CursorTokenTray", isDirectory: true)
        try FileManager.default.createDirectory(at: legacy, withIntermediateDirectories: true)
        try "{\"ok\":true}".write(to: legacy.appendingPathComponent("config.json"), atomically: true, encoding: .utf8)
        defer { try? FileManager.default.removeItem(at: home) }

        let dest = AppPaths.prepareDefaultDirectory(home: home)
        XCTAssertEqual(dest.lastPathComponent, "CursorRemain")
        XCTAssertTrue(FileManager.default.fileExists(atPath: dest.appendingPathComponent("config.json").path))
        XCTAssertEqual(AppPaths.appSupportName, "CursorRemain")
        XCTAssertEqual(AppPaths.displayName, "Cursor 余量")
    }
}

File: windows/CursorTokenCore.Tests/AppPathsTests.cs
using CursorTokenCore;
using Xunit;

namespace CursorTokenCore.Tests;

public class AppPathsTests
{
    [Fact]
    public void MigratesLegacyDirectoryWhenDestMissing()
    {
        var root = Path.Combine(Path.GetTempPath(), "cremain-mig-" + Guid.NewGuid().ToString("N"));
        var source = Path.Combine(root, "CursorTokenTray");
        var dest = Path.Combine(root, "CursorRemain");
        Directory.CreateDirectory(source);
        File.WriteAllText(Path.Combine(source, "config.json"), "{\"refresh_interval_minutes\":9}");
        File.WriteAllText(Path.Combine(source, "usage_history.jsonl"), "x\n");
        try
        {
            Assert.True(AppPaths.TryMigrateLegacyDirectory(source, dest));
            Assert.True(File.Exists(Path.Combine(dest, "config.json")));
            Assert.True(File.Exists(Path.Combine(dest, "usage_history.jsonl")));
            Assert.False(Directory.Exists(source));
            Assert.False(AppPaths.TryMigrateLegacyDirectory(source, dest));
        }
        finally
        {
            try { Directory.Delete(root, true); } catch { }
        }
    }

    [Fact]
    public void LeavesDestAloneWhenItAlreadyHasConfig()
    {
        var root = Path.Combine(Path.GetTempPath(), "cremain-keep-" + Guid.NewGuid().ToString("N"));
        var source = Path.Combine(root, "CursorTokenTray");
        var dest = Path.Combine(root, "CursorRemain");
        Directory.CreateDirectory(source);
        Directory.CreateDirectory(dest);
        File.WriteAllText(Path.Combine(source, "config.json"), "{\"from\":\"old\"}");
        File.WriteAllText(Path.Combine(dest, "config.json"), "{\"from\":\"new\"}");
        try
        {
            Assert.False(AppPaths.TryMigrateLegacyDirectory(source, dest));
            Assert.Equal("{\"from\":\"new\"}", File.ReadAllText(Path.Combine(dest, "config.json")));
            Assert.True(File.Exists(Path.Combine(source, "config.json")));
        }
        finally
        {
            try { Directory.Delete(root, true); } catch { }
        }
    }

    [Fact]
    public void CopiesIntoExistingEmptyDest()
    {
        var root = Path.Combine(Path.GetTempPath(), "cremain-copy-" + Guid.NewGuid().ToString("N"));
        var source = Path.Combine(root, "CursorTokenTray");
        var dest = Path.Combine(root, "CursorRemain");
        Directory.CreateDirectory(source);
        Directory.CreateDirectory(dest);
        File.WriteAllText(Path.Combine(source, "config.json"), "{\"ok\":true}");
        try
        {
            Assert.True(AppPaths.TryMigrateLegacyDirectory(source, dest));
            Assert.True(File.Exists(Path.Combine(dest, "config.json")));
            Assert.True(File.Exists(Path.Combine(source, "config.json")));
        }
        finally
        {
            try { Directory.Delete(root, true); } catch { }
        }
    }

    [Fact]
    public void OverrideDirectorySkipsDefaultName()
    {
        var dir = Path.Combine(Path.GetTempPath(), "cremain-ovr-" + Guid.NewGuid().ToString("N"));
        Assert.Equal(dir, AppPaths.ConfigDirectory(dir));
        Assert.Equal(AppPaths.AppName, "CursorRemain");
        Assert.Equal(AppPaths.LegacyAppName, "CursorTokenTray");
        Assert.Equal("Cursor 余量", AppPaths.DisplayName);
    }
}
